src/Microsoft.Azure.ServiceBus.MessageId/MessageIdPlugin.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.Azure.ServiceBus.Core;
+
+namespace Microsoft.Azure.ServiceBus.MessageId
+{
+    /// <summary>Generate Azure Service Bus <see cref="Message"/> id.</summary>
+    public class MessageIdPlugin : ServiceBusPlugin
+    {
+        private readonly Func<string> messageIdGenerator;
+
+        /// <summary>
+        /// <inheritdoc cref="Name"/>
+        /// </summary>
+        public override string Name => "Microsoft.Azure.ServiceBus.MessageId";
+
+        /// <summary>
+        /// Create a new instance of <see cref="MessageIdPlugin"/>
+        /// </summary>
+        /// <param name="messageIdGenerator"></param>
+        public MessageIdPlugin(Func<string> messageIdGenerator)
+        {
+            this.messageIdGenerator = messageIdGenerator;
+        }
+
+        /// <summary>
+        /// Assign message id if it's not already assigned.
+        /// <param name="message">The <see cref="Message"/> to assign id to.</param>
+        /// </summary>
+        /// <returns><see cref="Message"/> with id.</returns>
+        public override Task<Message> BeforeMessageSend(Message message)
```

Comment:
@nemakam thinking about it, this would be better handled via a separate issue. That way we can also encourage some community participation (fairly low barrier for engagement). Not to mention that issue would provide the context for future.